### PR TITLE
Add generic F103Vx targets

### PR DESCRIPTION
**Pull request by MCUdude**
This commit adds support for STM32F103V8, VB, VC, VD, VE, VF, and VG.
This is a pinout that's been requested in #722.

### CODE REVIEW

**Comment by fpistm** [comment]
Thanks @MCUdude

Linked to #958

**Comment by MCUdude** [comment]
BTW do you know where I can find pre-built linker scripts for other targets? Let's say I'd want to create a new pinout for a new target that's not currently supported in this core. Would I have to modify an existing script, or do you have a repo with a bunch of pre-built ones?

**Comment by fpistm** [comment]
Thanks @MCUdude 

You can use CubeMX to generate a project for STM32CubeIDE then you will find in the generated project the linker script and also the SystemClockConfig.

File: variants/Generic_F103Vx/PeripheralPins.c [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+  // {PA_0,  TIM5,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 1, 0)}, // TIM5_CH1
+#endif
+  {PA_1,  TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 2, 0)}, // TIM2_CH2
+  // {PA_1,  TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_TIM2_PARTIAL_2, 2, 0)}, // TIM2_CH2
+#if defined(STM32F103xE) || defined(STM32F103xG)
+  // {PA_1,  TIM5,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 2, 0)}, // TIM5_CH2
+#endif
+  {PA_2,  TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 3, 0)}, // TIM2_CH3
+  // {PA_2,  TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_TIM2_PARTIAL_1, 3, 0)}, // TIM2_CH3
+#if defined(STM32F103xG)
```
Comment (fpistm) [inline variants/Generic_F103Vx/PeripheralPins.c:135]:
```suggestion
#if defined(STM32F103xE) || defined(STM32F103xG)
```

File: variants/Generic_F103Vx/PeripheralPins.c [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+  // {PA_2,  TIM5,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 3, 0)}, // TIM5_CH3
+#endif
+#if defined(STM32F103xE) || defined(STM32F103xG)
```
Comment (fpistm) [inline variants/Generic_F103Vx/PeripheralPins.c:138]:
```suggestion
#ifdef STM32F103xG
```

File: variants/Generic_F103Vx/PeripheralPins.c [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+  // {PC_6,  TIM8,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 1, 0)}, // TIM8_CH1
+#endif
+  {PC_7,  TIM3,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_TIM3_ENABLE, 2, 0)}, // TIM3_CH2
+#if defined(STM32F103xE) || defined(STM32F103xG)
+  // {PC_7,  TIM8,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 2, 0)}, // TIM8_CH2
+#endif
+#if defined(STM32F103xE) || defined(STM32F103xG)
+  {PC_8,  TIM8,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 3, 0)}, // TIM8_CH3
+#else
+  {PC_8,  TIM3,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_TIM3_ENABLE, 3, 0)}, // TIM3_CH3
```
Comment (fpistm) [inline variants/Generic_F103Vx/PeripheralPins.c:223]: It is also available for xB

File: variants/Generic_F103Vx/PeripheralPins.c [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+#endif
+  {PA_6,  TIM3,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 1, 0)}, // TIM3_CH1
+#if defined(STM32F103xG)
+  // {PA_6,  TIM13,  STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 1, 0)}, // TIM13_CH1
+#endif
+  // {PA_7,  TIM3,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 2, 0)}, // TIM3_CH2
+#if defined(STM32F103xE) || defined(STM32F103xG)
+  {PA_7,  TIM8,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 1, 1)}, // TIM8_CH1N
+#else
+  {PA_7,  TIM1,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_TIM1_PARTIAL, 1, 1)}, // TIM1_CH1N
```
Comment (fpistm) [inline variants/Generic_F103Vx/PeripheralPins.c:158]: It is also available for xB

File: variants/Generic_F103Vx/PeripheralPins.c [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+  // {PA_2,  TIM5,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 3, 0)}, // TIM5_CH3
+#endif
+#if defined(STM32F103xE) || defined(STM32F103xG)
+  // {PA_2,  TIM9,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 1, 0)}, // TIM9_CH1
+#endif
+  // {PA_3,  TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_TIM2_PARTIAL_1, 4, 0)}, // TIM2_CH4
+#if defined(STM32F103xE) || defined(STM32F103xG)
+  {PA_3,  TIM5,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 4, 0)}, // TIM5_CH4
+#else
+  {PA_3,  TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE, 4, 0)}, // TIM2_CH4
```
Comment (fpistm) [inline variants/Generic_F103Vx/PeripheralPins.c:145]: It is also available for xB

File: variants/Generic_F103Vx/PeripheralPins.c [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+WEAK const PinMap PinMap_UART_TX[] = {
+  {PA_2,  USART2,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+  {PA_9,  USART1,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+  {PB_6,  USART1,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART1_ENABLE)},
+  {PB_10, USART3,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+#if defined(STM32F103xE) || defined(STM32F103xG)
+  {PC_10, UART4,   STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+#endif
+#if defined(STM32F103xB)
+  {PC_10, USART3,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART3_PARTIAL)},
```
Comment (fpistm) [inline variants/Generic_F103Vx/PeripheralPins.c:260]: It is also available for xB

File: variants/Generic_F103Vx/PeripheralPins.c [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+WEAK const PinMap PinMap_UART_RX[] = {
+  {PA_3,  USART2,  STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_NONE)},
+  {PA_10, USART1,  STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_NONE)},
+  {PB_7,  USART1,  STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_USART1_ENABLE)},
+  {PB_11, USART3,  STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_NONE)},
+#if defined(STM32F103xE) || defined(STM32F103xG)
+  {PC_11, UART4,   STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_NONE)},
+#endif
+#if defined(STM32F103xB)
+  {PC_11, USART3,  STM_PIN_DATA(STM_MODE_INPUT, GPIO_PULLUP, AFIO_USART3_PARTIAL)},
```
Comment (fpistm) [inline variants/Generic_F103Vx/PeripheralPins.c:281]: It is also available for xB

File: variants/Generic_F103Vx/PeripheralPins.c [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+
+//*** SPI ***
+
+#ifdef HAL_SPI_MODULE_ENABLED
+WEAK const PinMap PinMap_SPI_MOSI[] = {
+  {PA_7,  SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+#if defined(STM32F103xE) || defined(STM32F103xG)
+  {PB_5,  SPI3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+#else
+  {PB_5,  SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_SPI1_ENABLE)},
```
Comment (fpistm) [inline variants/Generic_F103Vx/PeripheralPins.c:324]: It is also available for xB

File: variants/Generic_F103Vx/PeripheralPins.c [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+};
+#endif
+
+#ifdef HAL_SPI_MODULE_ENABLED
+WEAK const PinMap PinMap_SPI_MISO[] = {
+  {PA_6,  SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+#if defined(STM32F103xE) || defined(STM32F103xG)
+  {PB_4,  SPI3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+#else
+  {PB_4,  SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_SPI1_ENABLE)},
```
Comment (fpistm) [inline variants/Generic_F103Vx/PeripheralPins.c:337]: It is also available for xB

File: variants/Generic_F103Vx/PeripheralPins.c [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+};
+#endif
+
+#ifdef HAL_SPI_MODULE_ENABLED
+WEAK const PinMap PinMap_SPI_SCLK[] = {
+  {PA_5,  SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+#if defined(STM32F103xE) || defined(STM32F103xG)
+  {PB_3,  SPI3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+#else
+  {PB_3,  SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_SPI1_ENABLE)},
```
Comment (fpistm) [inline variants/Generic_F103Vx/PeripheralPins.c:350]: It is also available for xB

File: variants/Generic_F103Vx/PeripheralPins.c [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+};
+#endif
+
+#ifdef HAL_SPI_MODULE_ENABLED
+WEAK const PinMap PinMap_SPI_SSEL[] = {
+  {PA_4,  SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+#if defined(STM32F103xE) || defined(STM32F103xG)
+  {PA_15, SPI3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
+#else
+  {PA_15, SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_SPI1_ENABLE)},
```
Comment (fpistm) [inline variants/Generic_F103Vx/PeripheralPins.c:363]: It is also available for xB

File: variants/Generic_F103Vx/PeripheralPins.c (hunk 6 above)
Comment (MCUdude) [reply variants/Generic_F103Vx/PeripheralPins.c:260]: No, UART4 on PC10 is not available for F103V8-B.

File: variants/Generic_F103Vx/PeripheralPins.c (hunk 7 above)
Comment (MCUdude) [reply variants/Generic_F103Vx/PeripheralPins.c:281]: No, UART4 on PC11 is not available for F103V8-B.

File: variants/Generic_F103Vx/PeripheralPins.c (hunk 8 above)
Comment (MCUdude) [reply variants/Generic_F103Vx/PeripheralPins.c:324]: F103V8-B does not have SPI3. This macro assigns SPI3 to PB5 on chips that have SPI3. F103V8-B uses SPI1 instead.

File: variants/Generic_F103Vx/PeripheralPins.c (hunk 9 above)
Comment (MCUdude) [reply variants/Generic_F103Vx/PeripheralPins.c:337]: F103V8-B does not have SPI3. This macro assigns SPI3 to PB4 on chips that have SPI3. F103V8-B uses SPI1 instead

File: variants/Generic_F103Vx/PeripheralPins.c (hunk 10 above)
Comment (MCUdude) [reply variants/Generic_F103Vx/PeripheralPins.c:350]: Same as above

File: variants/Generic_F103Vx/PeripheralPins.c (hunk 11 above)
Comment (MCUdude) [reply variants/Generic_F103Vx/PeripheralPins.c:363]: Same as above

File: variants/Generic_F103Vx/PeripheralPins.c (hunk 8 above)
Comment (MCUdude) [reply variants/Generic_F103Vx/PeripheralPins.c:324]: I could perhaps have written it like this:

```c
#ifdef HAL_SPI_MODULE_ENABLED
WEAK const PinMap PinMap_SPI_MOSI[] = {
  {PA_7,  SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
#if defined(STM32F103xE) || defined(STM32F103xG)
  // {PB_5,  SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_SPI1_ENABLE)},
  {PB_5,  SPI3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
#else
  {PB_5,  SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_SPI1_ENABLE)},
#endif
  {PB_15, SPI2, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)},
  {NC,    NP,    0}
};
#endif
```

**Comment by MCUdude** [comment]
OK, how about now?

File: variants/Generic_F103Vx/PeripheralPins.c (hunk 8 above)
Comment (fpistm) [reply variants/Generic_F103Vx/PeripheralPins.c:320]: Right, my comment was wrong, I would told the PB_5,  SPI1... is also available for other than xB

Comment (MCUdude) [reply variants/Generic_F103Vx/PeripheralPins.c:320]: I force-pushed a fix yesterday, where this has been solved

**Comment by TianYujiaEureka** [comment]
unsubscribe




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Hans"<notifications@github.com&gt;; 
发送时间: 2020年4月23日(星期四) 下午3:15
收件人: "stm32duino/Arduino_Core_STM32"<Arduino_Core_STM32@noreply.github.com&gt;; 
抄送: "Subscribed"<subscribed@noreply.github.com&gt;; 
主题: Re: [stm32duino/Arduino_Core_STM32] Add generic F103Vx targets (#1042)





 
@MCUdude commented on this pull request.
 
 
In variants/Generic_F103Vx/PeripheralPins.c:
 &gt; +  {PB_13, USART3,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART3_PARTIAL)}, +  {PD_3,  USART2,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART2_ENABLE)}, +  {PD_11, USART3,  STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_USART3_ENABLE)}, +  {NC,    NP,    0} +}; +#endif + +//*** SPI *** + +#ifdef HAL_SPI_MODULE_ENABLED +WEAK const PinMap PinMap_SPI_MOSI[] = { +  {PA_7,  SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)}, +#if defined(STM32F103xE) || defined(STM32F103xG) +  {PB_5,  SPI3, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_NONE)}, +#else +  {PB_5,  SPI1, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, AFIO_SPI1_ENABLE)},  
I force-pushed a fix yesterday, where this has been solved

**Pull request merged**